react-backend/client/src/components/admin/adminFeedbackPage.js

```diff
@@ -0,0 +1,158 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { Doughnut } from 'react-chartjs-2';
+import 'react-dates/initialize';
+import 'react-dates/lib/css/_datepicker.css';
+import { DateRangePicker } from 'react-dates';
+import { PrimaryButtonNoFill, PrimaryButton } from '../../shared/components/form/base';
+import { Heading, Link } from '../../shared/components/text/base';
+
+import Logoutbutton from '../visitors/logoutbutton';
+import { CbAdmin } from '../../api';
+
+const feedbackColors = [
+  {
+    feedback_score: -1,
+    label: 'Unimpressed',
+    backgroundColor: '#833FF7',
+    hoverBackgroundColor: '#6717F3',
+  },
+  {
+    feedback_score: 0,
+    label: 'Neutral',
+    backgroundColor: '#DBDBDB',
+    hoverBackgroundColor: '#666666',
+  },
+  {
+    feedback_score: 1,
+    label: 'Impressed',
+    backgroundColor: '#FDBD2D',
+    hoverBackgroundColor: '#DE9B06',
+  },
+];
+
+const donofig = (colorConfig, data) =>
+  ({
+    labels: feedbackColors.map(el => el.label),
+    datasets: [
+      {
+        data: feedbackColors.map(
+          el => data.filter(dat => dat.feedback_score === el.feedback_score)[0].count,
+        ),
+        backgroundColor: feedbackColors.map(el => el.backgroundColor),
+        hoverBackgroundColor: feedbackColors.map(el => el.hoverBackgroundColor),
+      },
+    ],
+  })
+;
+const lastCallStates = {
+  ALL: 'ALL',
+  DATERANGEPICKER: 'DATERANGEPICKER',
+};
+export default class AdminFeedbackPage extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      data: null,
+      error: null,
+      startDate: null,
+      endDate: null,
+      focusedInput: null,
+      lastCall: null,
+    };
+  }
+  componentDidMount() {
+    this.handleGetFeedback();
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+
+    if (
+      prevState.focusedInput !== this.state.focusedInput
```

Comment:
Part of the reason I refactored these request checkers is to keep the code in one place rather than passing them in as callbacks to `setState`.

With the datepicker I am using the change in `focusedInput` like an event listener to kick off the request. From what I can gather from documentation (which isn't great) and gh issues, this is the expected use.
 

Date picker request checks:
```
      prevState.focusedInput !== this.state.focusedInput // checks that it is called by a change in date picker
      && !this.state.focusedInput // checks that datepicker is being closed rather than opened
      && this.state.startDate // checks startDate is set
      && this.state.endDate // checks endDate is set (as both are needed)
```
All request checks:
```
      prevState.lastCall === lastCallStates.DATERANGEPICKER
      && this.state.lastCall === lastCallStates.ALL
```
☝️  This check is avoiding you from making fetch requests by repeatedly clicking 'ALL'. It will only fire another all request after a date picker request has been made.

Reply:
You should use the [onClose](http://airbnb.io/react-dates/?selectedKind=DRP%20-%20Calendar%20Props&selectedStory=with%20onClose%20handler&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel) handler instead of relying on focus.